### PR TITLE
add canSkipItemFailure handler func

### PR DESCRIPTION
**Pull request by ryanfkeepers**
adds a new func to the exchange backup handler: canskipItemFailure. This interface allows any handler to evaluate the provided error and runtime config to decide whether that error is able to be marked as skipped instead of return a recoverable error as per standard.

---

#### Does this PR need a docs update or release note?

- [x] :no_entry: No

#### Type of change

- [x] :bug: Bugfix

#### Test Plan

- [x] :zap: Unit test
- [x] :green_heart: E2E

### CODE REVIEW

**Review by ashmrtn — COMMENTED** [review]
non-event handlers need expanded tests prior to merging

File: src/internal/m365/collection/exchange/events_backup.go [hunk 1]
```diff
@@ -52,3 +58,22 @@ func (h eventBackupHandler) NewContainerCache(
 		getter: h.ac,
 	}
 }
+
+func (h eventBackupHandler) CanSkipItemFailure(
+	err error,
+	resourceID, itemID string,
+	opts control.Options,
```
[the remark this answers is not in the document]
Comment (ryanfkeepers) [reply src/internal/m365/collection/exchange/events_backup.go:65]: I don't see this as a problem.  Options are intended as a runtime configuration.  It feels natural to pass them around as needed.  Is your concern around passing in options vs the 503 map?  Or something else?

File: src/internal/m365/collection/exchange/contacts_backup_test.go [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+func TestContactsBackupHandlerUnitSuite(t *testing.T) {
+	suite.Run(t, &ContactsBackupHandlerUnitSuite{Suite: tester.NewUnitSuite(t)})
+}
+
+func (suite *ContactsBackupHandlerUnitSuite) TestHandler_CanSkipItemFailure() {
+	t := suite.T()
+
+	h := newContactBackupHandler(api.Client{})
+	cause, result := h.CanSkipItemFailure(nil, "", "", control.Options{})
```
[the remark this answers is not in the document]
Comment (ryanfkeepers) [reply src/internal/m365/collection/exchange/contacts_backup_test.go:27]: The tests for events aren't generically applicable for mail and contacts.  Primarily because the configuration that the tests depend on is also targeted to calendar events.  I could write in nil and non-nil checks with arbitrary data, but that seems like moot effort given the flat return of the func.  If anything, I kinda prefer the singular test, as evidence that we aren't considering any difference in inputs for those handlers; the output is always, simply, false.

File: src/internal/m365/collection/exchange/contacts_backup.go [hunk 3]
```diff
@@ -52,3 +54,11 @@ func (h contactBackupHandler) NewContainerCache(
 		getter: h.ac,
 	}
 }
+
+func (h contactBackupHandler) CanSkipItemFailure(
+	error,
+	string, string,
```
[the remark this answers is not in the document]
Comment (ryanfkeepers) [reply src/internal/m365/collection/exchange/contacts_backup.go:60]: yeah, totally fair

File: src/internal/m365/collection/exchange/events_backup_test.go [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+	table := []struct {
+		name        string
+		err         error
+		opts        control.Options
+		expect      assert.BoolAssertionFunc
+		expectCause fault.SkipCause
+	}{
+		{
+			name:   "no config",
+			err:    nil,
```
[the remark this answers is not in the document]
Comment (ryanfkeepers) [reply src/internal/m365/collection/exchange/events_backup_test.go:40]: Ah, I forgot to update that after swapping around some conditionals.  I'll make sure it gets an error when I plug in that data with the last PR.

File: src/internal/m365/collection/exchange/events_backup_test.go [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+			opts:   control.Options{},
+			expect: assert.False,
+		},
+		{
+			name: "empty skip on 503",
+			err:  nil,
```
[the remark this answers is not in the document]
Comment (ryanfkeepers) [reply src/internal/m365/collection/exchange/events_backup_test.go:46]: yarp

File: src/internal/m365/collection/exchange/events_backup.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (ashmrtn) [reply src/internal/m365/collection/exchange/events_backup.go:65]: my concern is that we're passing around all the options everywhere. This makes it tempting to access some of the info where we shouldn't and has led to some situations that could have strange or surprising interfaces (e.x. page size for enumeration is in options and passed to the `api.Client` code, but if we ever wanted to have fine grained control over page size would we end up ignoring the value in the client's options and instead take a parameter in the function call?)

File: src/internal/m365/collection/exchange/events_backup_test.go [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+				},
+			},
+			expect: assert.False,
+		},
+		{
+			name: "non-matching resource",
+			err:  clues.New("fix me I'm wrong"),
+			opts: control.Options{
+				SkipTheseEventsOnInstance503: map[string][]string{
+					"foo": []string{"bar", "baz"},
```
[the remark this answers is not in the document]
Comment (ryanfkeepers) [reply src/internal/m365/collection/exchange/events_backup_test.go:67]: I'll just patch in the item id on this one.  It's a moot point to have the current test if I also have the other.

File: src/internal/m365/collection/exchange/contacts_backup_test.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (ashmrtn) [reply src/internal/m365/collection/exchange/contacts_backup_test.go:27]: > The tests for events aren't generically applicable for mail and contacts

The new functions you added are still going to be called during backup because the handler is passed in as the skip checker in later PRs. That makes their implementation just as important as the implementation of the events handler. Part of writing code and tests is making sure the _wrong_ behavior doesn't happen. It can sometimes be just as much of the work as making sure the _right_ behavior does happen

This branch [dnm-bug-1](https://github.com/alcionai/corso/tree/dnm-bug-1) shows that unit tests will pass for contacts and mail _even if the implementation will lead to bugs in an actual backup_. What is being tested in this PR is "the implementation works right conditioned on the fact that the implementation is right" not that the implementation is correct no matter what it is

[the remark this answers is not in the document]
Comment (ryanfkeepers) [reply src/internal/m365/collection/exchange/contacts_backup_test.go:27]: To be fair, I think you're pointing out the ease of producing these conditions.  Not that it's possible.  The statement of possibility is a strawman argument, because nearly any function can be manufactured to introduce bugs while satisfying tests.  But that was particularly easy to do, I'll give you that.

[the remark this answers is not in the document]
Comment (ashmrtn) [reply src/internal/m365/collection/exchange/contacts_backup_test.go]: I think we may have different views of what tests are supposed to represent. In my mind tests should not have any knowledge of the implementation unless they're checking some optimizations. They should be blackbox tests that check for a given set of inputs the correct outputs for the component in question are returned. This requires having a spec of the expected behavior. In the case of mail and contacts, the spec should be "always return false when `CanSkipItemFailure` is called". However, a test with a single test case is not enough to show that for all inputs it returns false. Now, we can't really test "all inputs", so instead we choose representative sets of inputs, like the ones for the events test

Having tests that encode the spec expectations also helps guard against regressions in the future. The example of a bug I gave is a simple one that could be caught in review, if it's not overlooked because it's grouped together with a lot of other changes. There's a lot of other bugs that are equally dangerous that are much less obvious because they deal with subtle interactions between things, [like an error message changing](https://github.com/alcionai/corso/blob/5e8407a9702368b80c6d856b7ba309561a38e445/src/pkg/services/m365/api/mail.go#L588) or something. If these reside on code executed for edge cases and there's no tests for it then it could take a really long time to catch these issues

File: src/internal/m365/collection/exchange/contacts_backup_test.go [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+		expectCause fault.SkipCause
+	}{
+		{
+			name:   "no config",
+			err:    assert.AnError,
+			opts:   control.Options{},
+			expect: assert.False,
+		},
+		{
+			name: "empty skip on 503",
```
Comment (ashmrtn) [inline src/internal/m365/collection/exchange/contacts_backup_test.go:44]: nit: could you please fixup test names in both contacts and mail since these are all expected to return false?

**Review by ashmrtn — APPROVED** [review]

File: src/internal/m365/collection/exchange/contacts_backup_test.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (ryanfkeepers) [reply src/internal/m365/collection/exchange/contacts_backup_test.go]: It's true we differ on these things.  Admittedly, though, i could probably shift a bit further adapting for that viewpoint and be better in life as a result.
